python/xir/program.py

```diff
@@ -1,12 +1,14 @@
+"""This module contains the XIRProgram class and classes for the Xanadu IR"""
+
 import re
 import warnings
 from decimal import Decimal
-from typing import Any, Dict, List, Sequence, Set, Tuple, Union
+from typing import Any, Dict, Iterator, List, Mapping, Sequence, Tuple, Union
 
 from .decimal_complex import DecimalComplex
 from .utils import strip
 
-"""This module contains the XIRProgram class and classes for the Xanadu IR"""
+Wire = Union[int, str]
```

Comment:
Good question! I don't think the XIR currently supports float wire labels.  As a workaround, we could simply cast the float wire labels to strings in the PennyLane → XIR serializer although this is not ideal.  What are your thoughts, @thisac?

Reply:
We could just add support for writing wires as floats in the XIR if we want to have that. Otherwise, if we think it's reasonable to restrict what wires can be in the XIR (e.g. just integers and strings, which I find somewhat reasonable), casting all wires that are not those types in PennyLane to a string might be a reasonable approach. 

An issue with PennyLane is that it's not only restricted to using ints, floats or string either. It could be anything, e.g. a module or a function, so we probably need to restrict it somewhere. Perhaps attempting to cast to string if it's a float, complex, etc. and, if that doesn't work, use the hash value (which is also how PennyLane distinguishes wires internally).

Reply:
Thanks, Theo!  Those are some excellent points. One immediate question I have (perhaps @trbromley can speak to this) is why PennyLane bothers supporting `float`, `function`, etc. wire labels. What is the intended use case for this feature?

For the record, I would be OK with limiting the wire labels in the XIR to a "reasonable" set of types such as `Union[int, str]` and hashing or casting the rest to strings.

Reply:
The idea of the [Wires](https://pennylane.readthedocs.io/en/stable/code/api/pennylane.wires.Wires.html) class is both for internal validation and also to give some flexibility for custom wire labels such as ["Alice", "Bob"]. For me, the main use cases are ints and strings, with floats a rare but possible occurrence. Anything else I would consider unlikely and not something we'd have to support with the IR

Reply:
True, but what I'm struggling to understand is this, "rare but possible occurrence" for using `float` wire labels.  What is an example of a concrete scenario where using a float is preferred to (or, at least, on par with) using an integer or string?

Reply:
Agreed, I cannot think of a practical scenario where I'd rather call a wire `3.14` rather than `3` or `"some_name"`.

On the other hand, since the wires class is so flexible (maybe too flexible, at least from an IR perspective), we should think about what edge cases may occur outside of ints and strings. In this case - for me floats are the most likely edge-case I'd expect a user may try. Though the suggestion to cast to strings should be good!

Reply:
Sounds like a plan!  I've added a link to this discussion in the relevant story.